irc/client.go

```diff
@@ -0,0 +1,147 @@
+package irc
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"net"
+)
+
+type clientState int
+
+const (
+	clientStateRegistering clientState = iota
+	clientStateRegistered
+)
+
+type clientConnection struct {
+	conn   *net.TCPConn
+	config *Config
+
+	clientUser User
+	serverUser *User
+
+	state clientState
+
+	outgoingMessages chan *Message
+	shutdown         chan interface{}
+}
+
+func newClientConnection(conn *net.TCPConn, user *User, config *Config) *clientConnection {
+	ip, _, _ := net.SplitHostPort(conn.RemoteAddr().String())
+
+	return &clientConnection{
+		conn:   conn,
+		config: config,
+
+		clientUser: User{Nick: "*", Ident: user.Ident, Host: ip},
+		serverUser: user,
+
+		outgoingMessages: make(chan *Message),
+		shutdown:         make(chan interface{}),
+	}
+}
+
+func (cc *clientConnection) String() string {
+	return fmt.Sprintf("%v", cc.conn.RemoteAddr())
+}
+
+func (cc *clientConnection) handleConnInput() {
+	defer func() {
+		close(cc.shutdown)
+		cc.conn.Close()
+	}()
+
+	s := bufio.NewScanner(cc.conn)
+	for s.Scan() {
+		// Reads a line with s.Text() and parses it as
+		// an IRC message
+		msgStr := s.Text()
+		msg, err := StringToMessage(msgStr)
+
+		// Pretty sure error handling shouldn't be this complicated
+		if err != nil {
+			if err == ErrMalformedIRCMessage {
+				log.Printf("[%v] sent malformed IRC message: %v", cc, msgStr)
+			} else if numeric, ok := err.(*NumericReply); ok {
+				numeric.Target = cc.clientUser.Nick
+				fmt.Fprintln(cc.conn, numeric.ToMessage().String())
```

Comment:
I assume this is where we'll handle Numeric messages in the future (tm) ?

Reply:
numerics are only sent by server to client, so I don't think we'll need to, unless I'm missing something lmbo

Reply:
If they're only sent from server to client, we should never get them, right .. ? (Because we're a irc server and a slack client)

Reply:
yep.